plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_alerts/task.py

```diff
@@ -0,0 +1,381 @@
+        for index, alert in enumerate(alerts):
+
+            # todo - make method to convert
+            alert_time = alert.get(TIMESTAMP_KEY) / 1000
+            alert_time = datetime.fromtimestamp(alert_time, tz=timezone.utc).strftime(TIME_FORMAT)
+
+            if alert_time == start_time:
+                alert_hash = hash_sha1(alert)
+                if alert_hash not in old_hashes:
+                    deduped_alerts.append(alert)
+            elif alert_time > start_time:
+                deduped_alerts += alerts[index:]
+                print(f"{deduped_alerts= }")
+                break
+
+        num_alerts = len(deduped_alerts)
+        self.logger.info(f"Received {len(alerts)} alerts, and after dedupe there is {num_alerts} results.")
+
+        if deduped_alerts:
+            # alert results are already sorted by CS API, so get the latest timestamp
+            highest_timestamp = deduped_alerts[-1].get(TIMESTAMP_KEY)
+
+            for alert in deduped_alerts:
+                if alert.get(TIMESTAMP_KEY) == highest_timestamp:
+                    new_hashes.append(hash_sha1(alert))
+                    self.logger.info(f"Hashed latest event with timestamp {highest_timestamp}.")
+
+            self.logger.debug(f"Highest timestamp is {highest_timestamp}")
+            self.logger.debug(f"Last hash is {new_hashes}")
+
+        return deduped_alerts, new_hashes, highest_timestamp
+
+    @staticmethod
+    def _get_current_time():
+        return datetime.now(timezone.utc)
+
+    ###########################
+    # Custom Config
+    ###########################
+    def _parse_custom_config(
+        self, custom_config: Dict[str, Any], now: datetime, saved_state: Dict[str, str]
+    ) -> Tuple[str, int]:
+        """
+        Takes custom config from CPS and allows the specification of a new start time for alerts,
+        and allows the limit to be customised.
+
+        :param custom_config: custom_config for the plugin
+        :param now: datetime representing 'now'
+        :param saved_state: existing state of the integration
+        :return: string formatted time to apply to our alert queries, and limit integer to set how many
+        alerts to fetch per run.
+        """
+        state = saved_state.copy()
+
+        # set the alert limit from CPS if it exists, otherwise default to ALERT_LIMIT
+        alert_limit = custom_config.get("alert_limit", ALERT_LIMIT)
+
+        saved_time = state.get(LAST_ALERT_TIME)
+
+        log_msg = ""
+        if not saved_time:
+            log_msg += "No previous alert time within state. "
+            custom_timings = custom_config.get(LAST_ALERT_TIME, {})
+            custom_date = custom_timings.get("date")
+            custom_hours = custom_timings.get("hours", DEFAULT_LOOKBACK_HOURS)
+            start = datetime(**custom_date) if custom_date else (now - timedelta(hours=custom_hours))
+            state[LAST_ALERT_TIME] = start.strftime(TIME_FORMAT)
```

Comment:
I don't think we need to stringify our dates, a datetime object should be readable in logs and state and also easy to translate into epoch time for our queries.